### PR TITLE
Add postgres.EventBroadcaster

### PR DESCRIPTION
**Pull request by spooktheducks**

### CODE REVIEW

File: core/services/postgres/event_broadcaster.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+	defer b.listenersMu.Unlock()
+
+	for _, listener := range b.listeners {
+		err = multierr.Append(err, listener.stop())
+	}
+	b.listeners = nil // avoid "close of closed channel" panic on shutdown
+	return err
+}
+
+func (b *eventBroadcaster) Notify(channel string, payload interface{}) error {
```
Comment (se3000) [inline core/services/postgres/event_broadcaster.go:92]: Given that the payload is received as a string, I think it would make sense to make the payload a string here as well.

File: core/services/pipeline/orm_test.go [hunk 2]
```diff
@@ -238,8 +244,11 @@ func TestORM(t *testing.T) {
 
 			test := test
 			t.Run(test.name, func(t *testing.T) {
-				orm := pipeline.NewORM(db, config)
-				jobORM := job.NewORM(db, config, orm)
+				eventBroadcaster := postgres.NewEventBroadcaster(config.DatabaseURL(), 0, 0)
+				eventBroadcaster.Start()
+				defer eventBroadcaster.Stop()
+				orm := pipeline.NewORM(db, config, eventBroadcaster)
```
Comment (se3000) [inline core/services/pipeline/orm_test.go:250]:
```suggestion
				orm, cleanup := cltest.NewPipelineORM(db, config)
				defer cleanup()
```
I see these setup lines in a bunch of tests and I'm wondering if it'd be worth making a factory to cutdown on setup boilerplate?

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 3]
```diff
@@ -58,80 +56,67 @@ type ethBroadcaster struct {
...
 	trigger chan struct{}
 	chStop  chan struct{}
 	wg      sync.WaitGroup
+
+	utils.StartStopOnce
 }
 
 // NewEthBroadcaster returns a new concrete ethBroadcaster
-func NewEthBroadcaster(store *store.Store, config orm.ConfigReader) EthBroadcaster {
+func NewEthBroadcaster(store *store.Store, config orm.ConfigReader, eventBroadcaster postgres.EventBroadcaster) EthBroadcaster {
```
Comment (se3000) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:72]: I know we're trying to cut down the scope of what we inject. Would it be worth just passing in the subscription instead of passing around the whole event broadcaster?

File: core/services/bulletprooftxmanager/eth_broadcaster_test.go [hunk 4]
```diff
@@ -59,7 +60,11 @@ func TestEthBroadcaster_ProcessUnstartedEthTxs_Success(t *testing.T) {
 	ethClient := new(mocks.Client)
 	store.EthClient = ethClient
 
-	eb := bulletprooftxmanager.NewEthBroadcaster(store, config)
+	eventBroadcaster := postgres.NewEventBroadcaster(config.DatabaseURL(), 0, 0)
+	eventBroadcaster.Start()
+	defer eventBroadcaster.Stop()
+
+	eb := bulletprooftxmanager.NewEthBroadcaster(store, config, eventBroadcaster)
```
Comment (se3000) [inline core/services/bulletprooftxmanager/eth_broadcaster_test.go:67]:
```suggestion
	eb, cleanup := cltest.NewEthBroadcaster(store)
	defer cleanup()
```
Worth making a cltest factory for?

File: core/store/orm/schema.go [hunk 5]
```diff
@@ -23,6 +23,8 @@ type ConfigSchema struct {
 	ClientNodeURL                         string          `env:"CLIENT_NODE_URL" default:"http://localhost:6688"`
 	DatabaseTimeout                       models.Duration `env:"DATABASE_TIMEOUT" default:"500ms"`
 	DatabaseURL                           string          `env:"DATABASE_URL"`
+	DatabaseListenerMinReconnectInterval  time.Duration   `env:"DATABASE_LISTENER_MIN_RECONNECT_INTERVAL" default:"1m"`
```
Comment (se3000) [inline core/store/orm/schema.go:26]: `1m` naively seems high. What is the risk with trying to reconnect every 10 seconds?

File: core/services/job/mocks/spec.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery v2.2.1. DO NOT EDIT.
+// Code generated by mockery v2.1.0. DO NOT EDIT.
```
Comment (samsondav) [inline core/services/job/mocks/spec.go:1]: @spooktheducks this number probably shouldn't go backwards. Try updating your local mockery version?

File: core/services/job/spawner.go [hunk 7]
```diff
@@ -107,14 +113,14 @@ func (js *spawner) RegisterDelegate(delegate Delegate) {
...
 	defer close(js.chDone)
+	defer js.destroy()
 
 	// Initialize the Postgres event listener for new jobs
-	var chNewJobs <-chan string
-	listener, err := js.orm.ListenForNewJobs()
+	newJobs, err := js.orm.ListenForNewJobs()
 	if err != nil {
-		logger.Errorw("Job spawner failed to subscribe to 'new job' events, falling back to polling", "error", err)
+		logger.Warn("Job spawner could not subscribe to new job events, falling back to polling")
```
Comment (samsondav) [inline core/services/job/spawner.go:121]: Do we really want to limp on in this instance or give up and error out the whole application?

File: core/services/pipeline/orm.go [hunk 8]
```diff
@@ -277,41 +279,26 @@ func (o *orm) processNextUnclaimedTaskRun(ctx context.Context, fn ProcessTaskRun
...
 		}
-
 		return nil
 	})
-
 	if err != nil {
 		return errors.Wrap(err, "while processing task run")
 	}
 
+	// Emit a Postgres notification if this is the final `ResultTask`
```
Comment (samsondav) [inline core/services/pipeline/orm.go:289]: Why did you move this outside of the transaction?

File: core/services/pipeline/orm.go [hunk 9]
```diff
@@ -351,25 +338,18 @@ func (o *orm) AwaitRun(ctx context.Context, runID int64) error {
...
-		MinReconnectInterval: 1 * time.Second,
-		MaxReconnectDuration: 1 * time.Minute,
-	}
-	err := listener.Start()
+	sub, err := o.eventBroadcaster.Subscribe(postgres.ChannelRunCompleted, fmt.Sprintf("%d", runID))
 	if err != nil {
 		return err
 	}
-	defer logger.ErrorIfCalling(listener.Stop)
+	defer sub.Close()
```
Comment (samsondav) [inline core/services/pipeline/orm.go:345]: Does it need a logger.ErrorIfCalling ?

File: core/services/pipeline/orm_test.go (hunk 2 above)
Comment (samsondav) [reply core/services/pipeline/orm_test.go:250]: Agree with @se3000 about factory function

File: core/services/pipeline/runner.go [hunk 10]
```diff
@@ -75,23 +77,17 @@ func (r *runner) Stop() {
...
 func (r *runner) runLoop() {
 	defer close(r.chDone)
+	defer r.destroy()
 
-	var err error
-	r.newRunsListener, err = r.orm.ListenForNewRuns()
+	newRunsSubscription, err := r.orm.ListenForNewRuns()
 	if err != nil {
-		logger.Errorw(`Pipeline runner failed to subscribe to "new run" events, falling back to polling`, "error", err)
+		logger.Error("Pipeline runner could not subscribe to new run events, falling back to polling")
```
Comment (samsondav) [inline core/services/pipeline/runner.go:88]: Same here, is it even worth trying to continue if we couldnt even open the subscription?

File: core/services/postgres/advisory_lock.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+// NOTE: All advisory lock class IDs used by the Chainlink application MUST be
+// kept here to avoid accidental re-use
+const (
+	AdvisoryLockClassID_EthBroadcaster int32 = 0
+	AdvisoryLockClassID_JobSpawner     int32 = 1
+	AdvisoryLockClassID_EthConfirmer   int32 = 2
+
+	// ORM takes lock on 1027321974924625846 which splits into ClassID 239192036, ObjID 2840971190
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/postgres/advisory_lock.go:22]: Meaningless, I believe it was originally a randomly generated number.

PG advisory locks are actually always two 32-bit integers. If you give one 64-bit integer it will do some sort of bitshifting to break it into two 32-bit integers internally.

These two 32-bit ints are the numbers that postgres uses internally for this 64-bit lock ID. Found by actually querying the postgres lock table while the application advisory lock is being held.

File: core/services/postgres/advisory_lock.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	AdvisoryLockClassID_ORM int32 = 239192036
+
+	AdvisoryLockObjectID_EthConfirmer int32 = 0
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/postgres/advisory_lock.go:25]: This is an object ID, not a class ID

File: core/services/postgres/advisory_lock.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+type AdvisoryLock struct {
+	URI  string
+	conn *sql.Conn
+	db   *sql.DB
+	mu   *sync.Mutex
+}
+
+// TODO(sam): Fix this to be a single global connection not many individual ones
+// See: https://www.pivotaltracker.com/story/show/175288169
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/postgres/advisory_lock.go:36]: I believe spook addressed Event subscriptions in this PR, not the advisory locks. That would be additional and separate work.

File: core/services/postgres/channels.go [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package postgres
+
+const (
```
Comment (samsondav) [inline core/services/postgres/channels.go:3]: I like this, its neat :+1:

File: core/services/postgres/event_broadcaster.go [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+type eventBroadcaster struct {
+	uri                  string
+	minReconnectInterval time.Duration
+	maxReconnectDuration time.Duration
+	db                   *sql.DB
+
+	listeners   map[string]*channelListener
+	listenersMu sync.Mutex
```
Comment (samsondav) [inline core/services/postgres/event_broadcaster.go:36]: We are inconsistent whether we use a pointer or embed the mutex directly on the struct. Does it matter?

File: core/services/postgres/event_broadcaster.go (hunk 1 above)
Comment (samsondav) [inline core/services/postgres/event_broadcaster.go:92]: This is convenient, but the problem is we may wish to atomically generate notification events inside a transaction. In that case, the NOTIFY must be issued on the parent DB connection - not the eventBroadcaster's db connection.

File: core/services/postgres/event_broadcaster.go [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+	_, err := b.db.Exec(`SELECT pg_notify($1, $2::text)`, channel, payload)
+	return errors.Wrap(err, "Postgres event broadcaster could not notify")
+}
+
+func (b *eventBroadcaster) Subscribe(channel, payloadFilter string) (Subscription, error) {
+	b.listenersMu.Lock()
+	defer b.listenersMu.Unlock()
+
+	if b.listeners[channel] == nil {
+		listener := newChannelListener(b.uri, b.minReconnectInterval, b.maxReconnectDuration, channel, b)
```
Comment (samsondav) [inline core/services/postgres/event_broadcaster.go:102]: I may be misunderstanding, but doesn't this still open a new connection for every new subscription?

(Which is the situation we are trying to get away from)

File: core/services/postgres/transaction.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
+package postgres
```
Comment (samsondav) [inline core/services/postgres/transaction.go:1]: This package is getting much nicer

File: core/store/orm/config.go [hunk 18]
```diff
@@ -174,6 +174,14 @@ func (c Config) getDuration(s string) models.Duration {
 	return rv
 }
 
+func (c Config) DatabaseListenerMinReconnectInterval() time.Duration {
+	return c.viper.GetDuration(EnvVarName("DatabaseListenerMinReconnectInterval"))
```
Comment (samsondav) [inline core/store/orm/config.go:178]: It seems like we don't actually use this? It looks hardcoded to me

**Review by samsondav — DISMISSED** [review]

File: core/services/postgres/advisory_lock.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/advisory_lock.go:36]: Yeah, the AdvisoryLock and GormTransaction bits were just moved from `core/utils` to `core/services/postgres`.  Didn't want to add scope.

File: core/services/postgres/advisory_lock.go [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+	return multierr.Combine(connErr, dbErr)
+}
+
+func (lock *AdvisoryLock) TryLock(ctx context.Context, classID int32, objectID int32) (err error) {
+	lock.mu.Lock()
+	defer lock.mu.Unlock()
+	defer utils.WrapIfError(&err, "TryAdvisoryLock failed")
+
+	if lock.conn == nil {
+		db, err2 := sql.Open("postgres", lock.URI)
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/advisory_lock.go:75]: One of our static checks (shadow) didn't like me reusing `err`

File: core/services/postgres/event_broadcaster.go [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+
+		for {
+			select {
+			case <-listener.chStop:
+				return
+
+			case notification, open := <-listener.NotificationChannel():
+				if !open {
+					return
+				} else if notification == nil {
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:200]: I don't believe it ever actually returns nil, but I still wanted to guard against it (e.g. if the upstream package ends up with a bug at some point)

File: core/services/postgres/event_broadcaster.go [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+					continue
+				}
+				logger.Debugw("Postgres listener: received notification",
+					"channel", notification.Channel,
+					"payload", notification.Extra,
+				)
+				listener.broadcast(notification)
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:207]: The best approach would be to immediately put the notification in a queue (probably `utils.BoundedQueue`) and then kick off a `utils.SleeperTask`.  If that sounds good, I'll get it in today

File: core/services/postgres/event_broadcaster.go [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+
+	sub.close()
+	delete(listener.subscriptions, sub)
+}
+
+func (listener *channelListener) stop() error {
+	if !listener.OkayToStop() {
+		return errors.Errorf("Postgres event listener has already been stopped (channel: %v)", listener.channel)
+	}
+	err := listener.Listener.Close()
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:265]: Yeah, I didn't want the `pq.Listener` to get deadlocked trying to send to its subscribers, which seems possible if they stop listening before it stops sending.  Just playing active defense against upstream code here.  Happy to add a comment

File: core/services/postgres/event_broadcaster.go (hunk 1 above)
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:92]: Sure thing, works for me

File: core/services/bulletprooftxmanager/eth_broadcaster.go (hunk 3 above)
Comment (spooktheducks) [reply core/services/bulletprooftxmanager/eth_broadcaster.go:72]: We could, but I encountered two problems with that:
1) subscribing in application.go seems strange
2) `pipeline.ORM` has to create and destroy subscriptions continually so it needs the EventBroadcaster regardless

I'm curious about
> trying to cut down the scope of what we inject

I've been actively trying to increase the scope what we inject (the above is a pattern that's worked well for me in the past, particularly when it comes to testing).  Should we circle up on this?

File: core/store/orm/schema.go (hunk 5 above)
Comment (spooktheducks) [reply core/store/orm/schema.go:26]: These were just the defaults from the `pq` lib's examples.  Happy to change it but curious to hear @samsondav's take on performance before I do

File: core/services/job/spawner.go (hunk 7 above)
Comment (spooktheducks) [reply core/services/job/spawner.go:121]: I wouldn't call it limping, as it's still fully functional, just a little slower.  Not erroring out the whole application might make whatever issue is causing this easier to debug.

File: core/services/pipeline/runner.go (hunk 10 above)
Comment (spooktheducks) [reply core/services/pipeline/runner.go:88]: Same response as above.  Still works, just a little bit slower, and having a running node can help with debugging

File: core/services/postgres/advisory_lock.go [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+	lock.db = nil
+	lock.conn = nil
+
+	return multierr.Combine(connErr, dbErr)
+}
+
+func (lock *AdvisoryLock) TryLock(ctx context.Context, classID int32, objectID int32) (err error) {
+	lock.mu.Lock()
+	defer lock.mu.Unlock()
+	defer utils.WrapIfError(&err, "TryAdvisoryLock failed")
```
Comment (j16r) [inline core/services/postgres/advisory_lock.go:72]: I feel like the intention here is good but ends up obscuring the error in a way that literally writing out `errors.Wrap(err, "more context for this error at this particular line")` would, e.g:

```go

rows, err := lock.conn.QueryContext(ctx, "SELECT pg_try_advisory_lock($1, $2)", classID, objectID)
if err != nil {
  return errors.Wrap(err, "failed to select query lock")
}
```

File: core/services/postgres/transaction.go [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+)
+
+func GormTransaction(ctx context.Context, db *gorm.DB, fc func(tx *gorm.DB) error, txOptss ...sql.TxOptions) (err error) {
```
Comment (j16r) [inline core/services/postgres/transaction.go:11]: Go's SQL primitives are designed so you can always `defer rollback` and if the transaction was committed first it'll do nothing, so I'm not sure this helper function gets you all that much?

File: core/services/pipeline/orm.go (hunk 9 above)
Comment (spooktheducks) [reply core/services/pipeline/orm.go:345]: Doesn't return an error

File: core/services/postgres/event_broadcaster.go (hunk 15 above)
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:36]: It's probably best to not embed (anonymously) so that nothing external can try to `.Lock()` it.  Regarding pointer vs. non-pointer, given that it's a private field and nothing else can see it/take a reference to it, it really doesn't make a difference.  I like using pre-initialized zero values whenever possible, which `sync.Mutex` was designed for.

File: core/services/postgres/event_broadcaster.go (hunk 16 above)
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:102]: It opens one per Postgres channel name (i.e., if the node is executing 10 jobs at once, it won't open 10 `pipeline_run_completed` connections, just one).  It didn't look like you could simply call `LISTEN` without arguments, so we're stuck with this.

File: core/services/postgres/transaction.go (hunk 24 above)
Comment (spooktheducks) [reply core/services/postgres/transaction.go:11]: You might be right... I was just moving this from the `utils` package, though, so it's probably out of scope for this PR to change up our approach.  I'll look into it and will add a new story if it seems like a good approach.

File: core/store/orm/config.go (hunk 18 above)
Comment (spooktheducks) [reply core/store/orm/config.go:178]: Nah, check `core/services/chainlink/application.go` line 106:
```go
eventBroadcaster := postgres.NewEventBroadcaster(config.DatabaseURL(), config.DatabaseListenerMinReconnectInterval(), config.DatabaseListenerMaxReconnectDuration())
```

File: core/services/postgres/advisory_lock.go (hunk 23 above)
Comment (spooktheducks) [reply core/services/postgres/advisory_lock.go:72]: FWIW, `utils.WrapIfError` passes through to `errors.Wrap`, which annotates errors with a stack trace.  how Go handles `defer` statements, that stack trace points to the exact line at which the `return err` happens, meaning that we get the precise info that you're referring to.

Comment (j16r) [reply core/services/postgres/advisory_lock.go:72]: Yeah I'm referring more to the English part of the error message.

File: core/services/postgres/advisory_lock.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/postgres/advisory_lock.go:22]: Followup to my comment above, I was mistaken.

The two numbers are actually nonsense :). Bigint keys operate on a completely different scope than the two 32-bit keys.

[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/postgres/advisory_lock.go:22]: This comment should be removed.

File: core/store/orm/schema.go (hunk 5 above)
Comment (samsondav) [reply core/store/orm/schema.go:26]: I don't have an opinion on this. Either 1m or 10s is fine by me.

File: core/services/postgres/event_broadcaster.go (hunk 15 above)
Comment (samsondav) [reply core/services/postgres/event_broadcaster.go:38]: I suppose I mean if you copy the struct, wouldn't it copy the mutex too?

File: core/services/postgres/event_broadcaster.go (hunk 16 above)
Comment (samsondav) [reply core/services/postgres/event_broadcaster.go:109]: You can (and we should) issue multiple `LISTEN` commands over a single connection.

Similarly if you unregister you can `UNLISTEN`.

The way this currently works, it will still create multiple connections for multiple types of notifications. We should attempt to get this down to one connection.

File: core/services/postgres/event_broadcaster.go (hunk 15 above)
Comment (spooktheducks) [reply core/services/postgres/event_broadcaster.go:38]: `sync.Mutex` has special handling -- it's disallowed from being copied.  I believe `go vet` is what surfaces the error when you try to do so

File: core/services/postgres/advisory_lock.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/postgres/advisory_lock.go:22]: Let's do it when we do the next batch of work on the advisory lock (i.e. converting `orm.PostgresLockingStrategy` to use `postgres.AdvisoryLock`

File: core/services/postgres/advisory_lock.go (hunk 23 above)
Comment (spooktheducks) [reply core/services/postgres/advisory_lock.go:72]: We're going to be gutting this entire type very soon in favor of `postgres.AdvisoryLock`.  I'll make a note in the tracker story to take care of this as well.

**Review by se3000 — APPROVED** [review]
